**Note: It will take around 15 mins to run. (and will definitely not crash if each sell is run individually)**

In [12]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import urllib.request
from urllib.request import urlopen
A = 'https://raw.githubusercontent.com/Samagra06/ML/main/Data_Q1.txt'


Reading the dataset in Data Q1.txt into two numpy arrays

In [13]:
from sklearn.datasets import load_svmlight_file
with urlopen(A) as f:
  X,Y = load_svmlight_file(f)

X = X.toarray()
print('Label',Y)
print('Features',X)  

Label [ 1. -1. -1. ...  1. -1.  1.]
Features [[1. 1. 1. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 1. 1. 1.]
 [0. 0. 0. ... 0. 0. 0.]]


Number of classes in the data set

In [14]:
unique_values = set(Y)
count = len(unique_values)
print("No of claases are", count)


No of claases are 2


No. of samples in each class

In [15]:
count_1 = np.count_nonzero(Y == 1)
print("No of Samples with label 1 = ", count_1)
count_2 = np.count_nonzero(Y == -1)
print("No of Samples with label -1 = ",count_2)

No of Samples with label 1 =  2210
No of Samples with label -1 =  1933


**Since the samples for the two classes donot differ by much, we can conclude that there is no class imbalance.**

In [16]:
X = pd.DataFrame(X)
Y = pd.DataFrame(Y)

In [17]:
X = X.astype('int32')
Y = Y.astype('int32')

Spliiting the dataset into test and train

In [18]:
X_train = X.sample(frac = 0.8, random_state = 200)
X_test = X.drop(X_train.index)
Y_train = Y.sample(frac = 0.8, random_state = 200)
Y_test = Y.drop(Y_train.index)

In [19]:
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

In [20]:
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import warnings
warnings.filterwarnings("ignore")

In [21]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix

**Training the dataset with different algorithms**

**i) Logistic Regression with L2 regulariser**

In [22]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(penalty='l2', solver='liblinear')
param_grid = {'C': [0.01, 0.1, 1, 10, 100]}
grid_search = GridSearchCV(lr, param_grid, cv=5)
grid_search.fit(X_train, Y_train)
best_model = grid_search.best_estimator_
Y_pred_train_LR_L2 = best_model.predict(X_train)
Y_pred_test_LR_L2 = best_model.predict(X_test)
print("Best hyperparameters: ", grid_search.best_params_)
print("Best score: ", grid_search.best_score_)
lr_nonzero = np.count_nonzero(best_model.coef_)
print('no. of non zero coefficients:', lr_nonzero)

Best hyperparameters:  {'C': 1}
Best score:  0.9007186048948977
no. of non zero coefficients: 48117


**ii) Logistic Regression with L1 regulariser**

In [23]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(penalty='l1', solver='liblinear')
param_grid = {'C': [0.01, 0.1, 1, 10, 100]}
grid_search = GridSearchCV(lr, param_grid, cv=5)
grid_search.fit(X_train, Y_train)
best_model = grid_search.best_estimator_
Y_pred_train_LR_L1 = best_model.predict(X_train)
Y_pred_test_LR_L1 = best_model.predict(X_test)
print("Best hyperparameters: ", grid_search.best_params_)
print("Best score: ", grid_search.best_score_)
lr_nonzero = np.count_nonzero(best_model.coef_)
print('no. of non zero coefficients:', lr_nonzero)

Best hyperparameters:  {'C': 1}
Best score:  0.8874456033866022
no. of non zero coefficients: 656


**iii) SVM with L2 regulariser**

In [24]:
kfold = KFold(n_splits=5, shuffle=True, random_state=42)
param_grid = {'C': [0.001, 0.01, 0.1, 1, 10, 100]}
model = LinearSVC(penalty='l2', dual=False)
grid_search = GridSearchCV(model, param_grid, cv=kfold, n_jobs=-1)
grid_search.fit(X_train, Y_train)
print("Best hyperparameters: ", grid_search.best_params_)
best_model = grid_search.best_estimator_
Y_pred_train_SVM_L1 = best_model.predict(X_train)
Y_pred_test_SVM_L1 = best_model.predict(X_test)
lr_nonzero = np.count_nonzero(best_model.coef_)
print('no. of non zero coefficients:', lr_nonzero)

Best hyperparameters:  {'C': 0.01}
no. of non zero coefficients: 48117


**iv) SVM with L1 regulariser**

In [25]:
kfold = KFold(n_splits=5, shuffle=True, random_state=42)
param_grid = {'C': [0.001, 0.01, 0.1, 1, 10, 100]}
model = LinearSVC(penalty='l1', dual=False)
grid_search = GridSearchCV(model, param_grid, cv=kfold, n_jobs=-1)
grid_search.fit(X_train, Y_train)
print("Best hyperparameters: ", grid_search.best_params_)
best_model = grid_search.best_estimator_
Y_pred_train_SVM_L2 = best_model.predict(X_train)
Y_pred_test_SVM_L2 = best_model.predict(X_test)
lr_nonzero = np.count_nonzero(best_model.coef_)
lr_nonzero

Best hyperparameters:  {'C': 100}


1590

In [26]:
                                                        # TAKING TOO MUCH TIME
# param_grid = {'gamma': [0.01, 0.1, 1, 10]}
# svc = SVC(kernel='rbf')
# grid_search = GridSearchCV(svc, param_grid, cv=5)
# grid_search.fit(X_train, Y_train)
# best_model = grid_search.best_estimator_
# Y_pred_train = best_model.predict(X_train)
# Y_pred_test = best_model.predict(X_test)

**v) KNN**

In [27]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()
param_grid = {'n_neighbors': [3, 5, 7]}
grid_search = GridSearchCV(knn, param_grid=param_grid, cv=5)
grid_search.fit(X_train, Y_train)
best_model = grid_search.best_estimator_
Y_pred_train_KNN = best_model.predict(X_train)
Y_pred_test_KNN = best_model.predict(X_test)

**vi) Decision Tree**

In [28]:
from sklearn.tree import DecisionTreeClassifier
param_grid = {'min_weight_fraction_leaf': [0.0, 0.1, 0.2, 0.3, 0.4, 0.5]}
dtc = DecisionTreeClassifier()
grid_search = GridSearchCV(dtc, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, Y_train)
print("Best hyperparameters: ", grid_search.best_params_)
best_model = grid_search.best_estimator_
Y_pred_train_DT = best_model.predict(X_train)
Y_pred_test_DT = best_model.predict(X_test)

Best hyperparameters:  {'min_weight_fraction_leaf': 0.0}


**vi) Random Forest**

In [29]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=42)
param_grid = {'n_estimators': [50, 100, 150]}
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, n_jobs=-1)
grid_search.fit(X_train, Y_train)
best_model = grid_search.best_estimator_
Y_pred_train_RF = best_model.predict(X_train)
Y_pred_test_RF = best_model.predict(X_test)
print("Best hyperparameters: ", grid_search.best_params_)

Best hyperparameters:  {'n_estimators': 50}


Defining a function for accuracy, precision, recall, specificity and senstivity values 

In [30]:
def Table(Y_train, Y_pred_train,Y_test, Y_pred_test):
  accuracy_train = accuracy_score(Y_train, Y_pred_train)
  accuracy_test = accuracy_score(Y_test, Y_pred_test)
  precision_train = precision_score(Y_train, Y_pred_train)
  precision_test = precision_score(Y_test, Y_pred_test)
  recall_train = recall_score(Y_train, Y_pred_train)
  recall_train = recall_score(Y_test, Y_pred_test)
  tn_train, fp_train, fn_train, tp_train = confusion_matrix(Y_train, Y_pred_train).ravel()
  tn_test, fp_test, fn_test, tp_test = confusion_matrix(Y_test, Y_pred_test).ravel()
  specificity_train = tn_train / (tn_train + fp_train)
  specificity_test = tn_test / (tn_test + fp_test)
  sensitivity_train = tp_train / (tp_train + fn_train)
  sensitivity_test = tp_test / (tp_test + fn_test)
  return accuracy_train, accuracy_test, precision_train, precision_test, recall_train, recall_train, specificity_train, specificity_test, sensitivity_train, sensitivity_test

Accuracy, precision, recall, specificity and senstivity values for different methods 

In [31]:
LR_L2_metrics = Table(Y_train, Y_pred_train_LR_L2, Y_test, Y_pred_test_LR_L2)
LR_L1_metrics = Table(Y_train, Y_pred_train_LR_L1, Y_test, Y_pred_test_LR_L1)
SVM_L2_metrics = Table(Y_train, Y_pred_train_SVM_L2, Y_test, Y_pred_test_SVM_L2)
SVM_L1_metrics = Table(Y_train, Y_pred_train_SVM_L1, Y_test, Y_pred_test_SVM_L1)
KNN_metrics = Table(Y_train, Y_pred_train_KNN, Y_test, Y_pred_test_KNN)
DT_metrics = Table(Y_train, Y_pred_train_DT, Y_test, Y_pred_test_DT)
RF_metrics = Table(Y_train, Y_pred_train_RF, Y_test, Y_pred_test_RF)
print(LR_L2_metrics)
print(LR_L1_metrics)
print(SVM_L2_metrics)
print(SVM_L1_metrics)
print(KNN_metrics)
print(DT_metrics)
print(RF_metrics)

(0.9981894990947495, 0.9034981905910736, 0.9983098591549295, 0.8867102396514162, 0.9356321839080459, 0.9356321839080459, 0.9980506822612085, 0.868020304568528, 0.9983098591549295, 0.9356321839080459)
(0.9891369945684972, 0.8866103739445115, 0.9854829704075935, 0.8763796909492274, 0.9126436781609195, 0.9126436781609195, 0.9831059129304743, 0.8578680203045685, 0.9943661971830986, 0.9126436781609195)
(0.9990947495473748, 0.8902291917973462, 0.9994363021420518, 0.8805309734513275, 0.9149425287356322, 0.9149425287356322, 0.9993502274204028, 0.8629441624365483, 0.9988732394366198, 0.9149425287356322)
(0.9900422450211225, 0.9131483715319663, 0.9876819708846585, 0.8903225806451613, 0.9517241379310345, 0.9517241379310345, 0.9857050032488629, 0.8705583756345178, 0.9938028169014085, 0.9517241379310345)
(0.8527459263729632, 0.781664656212304, 0.9735099337748344, 0.9319727891156463, 0.6298850574712643, 0.6298850574712643, 0.9766081871345029, 0.949238578680203, 0.7453521126760564, 0.6298850574712643

In [32]:
Array = [LR_L2_metrics,LR_L1_metrics, SVM_L2_metrics, SVM_L1_metrics, KNN_metrics, DT_metrics, RF_metrics ]

In [33]:
Array = np.round(Array, 3)
Array

array([[0.998, 0.903, 0.998, 0.887, 0.936, 0.936, 0.998, 0.868, 0.998,
        0.936],
       [0.989, 0.887, 0.985, 0.876, 0.913, 0.913, 0.983, 0.858, 0.994,
        0.913],
       [0.999, 0.89 , 0.999, 0.881, 0.915, 0.915, 0.999, 0.863, 0.999,
        0.915],
       [0.99 , 0.913, 0.988, 0.89 , 0.952, 0.952, 0.986, 0.871, 0.994,
        0.952],
       [0.853, 0.782, 0.974, 0.932, 0.63 , 0.63 , 0.977, 0.949, 0.745,
        0.63 ],
       [0.999, 0.882, 0.999, 0.867, 0.915, 0.915, 0.999, 0.845, 0.999,
        0.915],
       [0.999, 0.884, 0.999, 0.845, 0.954, 0.954, 0.999, 0.807, 0.999,
        0.954]])

**Tabulating the accuracy, precision, recall, specificity and senstivity values for training set and test set for each model**

In [34]:
Tab = pd.DataFrame(Array, columns = ['accuracy_train', 'accuracy_test', 'precision_train', 'precision_test', 'recall_train', 'recall_train', 'specificity_train', 'specificity_test', 'sensitivity_train', 'sensitivity_test'],index=['Log Reg L2', 'Log Reg L1', 'SVM L2', 'SVM L1', 'KNN', 'Decision Tree', 'Random Forest'])
Tab

accuracy_train  accuracy_test  precision_train  precision_test  \
Log Reg L2              0.998          0.903            0.998           0.887   
Log Reg L1              0.989          0.887            0.985           0.876   
SVM L2                  0.999          0.890            0.999           0.881   
SVM L1                  0.990          0.913            0.988           0.890   
KNN                     0.853          0.782            0.974           0.932   
Decision Tree           0.999          0.882            0.999           0.867   
Random Forest           0.999          0.884            0.999           0.845   

               recall_train  recall_train  specificity_train  \
Log Reg L2            0.936         0.936              0.998   
Log Reg L1            0.913         0.913              0.983   
SVM L2                0.915         0.915              0.999   
SVM L1                0.952         0.952              0.986   
KNN                   0.630         0.630              0.977   
Decision Tree         0.915         0.915              0.999   
Random Forest         0.954         0.954              0.999   

               specificity_test  sensitivity_train  sensitivity_test  
Log Reg L2                0.868              0.998             0.936  
Log Reg L1                0.858              0.994             0.913  
SVM L2                    0.863              0.999             0.915  
SVM L1                    0.871              0.994             0.952  
KNN                       0.949              0.745             0.630  
Decision Tree             0.845              0.999             0.915  
Random Forest             0.807              0.999             0.954

Comparison of sparcity of L1 and L2 regulariser

**As we can see that the no. of non-zero coefficients obtained in logistic Regression and SVM using L1 regulariser is 1878 and 1559 respectively, whereas in the case of L2 regulariser, the value is 48117 which is significantly higher and this is sufficient t oprove that L1 regularization encourages sparsity in the model, while L2 regularization does not necessarily lead to sparse models.**

**This reason is because L1 regularization tends to force some coefficients to zero, effectively removing them from the model and resulting in a sparse model.**